### PR TITLE
Fetch: set-cookie special cases in Headers

### PR DESCRIPTION
**Pull request by lucacasonato**
**Do not merge**

Spec change: https://github.com/whatwg/fetch/pull/1346

### CODE REVIEW

File: fetch/api/headers/header-setcookie.any.js
Comment (annevk) [inline fetch/api/headers/header-setcookie.any.js:33]: I'd like to see more iterator tests, including for `set-cookie2` and sorting.

Comment (lucacasonato) [reply fetch/api/headers/header-setcookie.any.js:41]: Done.

**Review by annevk — COMMENTED** [review]
These look great. One ordering case that isn't covered yet I think is something like:

* Append (Set-Cookie, 1)
* Append (A-Cool-Header, 2)
* Append (Set-Cookie, 3)
* Append (A-Cool-Header, 4)

and perhaps variants on that. As you implemented this you probably know more what could go wrong. Thoughts?

**Comment by lucacasonato** [comment]
Added a rather intricate ordering case that ensures that keys are ordered alphabetically, but the value order is never touched (for both set-cookie, and all other headers).

I took a look through the Deno unit tests for `Header` that touch `set-cookie`, and I think this PR covers all cases. (grep for `set-cookie` in https://github.com/denoland/deno/blob/3f0cf9619fce71a8898c495501df4bdb0e07e735/cli/tests/unit/headers_test.ts).

**Review by annevk — APPROVED** [review]
Excellent, thanks!

**Comment by lucacasonato** [comment]
Just ran the tests against Deno, and found some issues: `assert_array_equals` is not a deep equals it seems, so we get errors like these:

```
Headers iterator does not combine set-cookie headers
assert_array_equals: expected property 0 to be ["set-cookie", "foo=bar"] but got ["set-cookie", "foo=bar"] (expected array [["set-cookie", "foo=bar"], ["set-cookie", "fizz=buzz; domain=example.com"]] got [["set-cookie", "foo=bar"], ["set-cookie", "fizz=buzz; domain=example.com"]])
    at Test.<anonymous> (http://web-platform.test:8000/fetch/api/headers/header-setcookie.any.js:37:3)
    at Test.step (http://web-platform.test:8000/resources/testharness.js:2087:25)
    at test (http://web-platform.test:8000/resources/testharness.js:557:30)
    at http://web-platform.test:8000/fetch/api/headers/header-setcookie.any.js:34:1
    at importScripts (deno:runtime/js/99_main.js:197:24)
```

It is checking reference equality not value equality... 

Is there a deep equals assertion thing, or am I going to have to roll my own?

**Comment by lucacasonato** [comment]
Fixed now. All tests (with the exception of the getSetCookie tests, because that function is not implemented), pass now.

File: fetch/api/headers/header-setcookie.any.js
Comment (jimmywarting) [inline fetch/api/headers/header-setcookie.any.js:158]: Guess this should be:
```js
const headers = new Headers([["set-cookie", ""]]);
// or: 
const headers = new Headers({"set-cookie": ""});
```

File: fetch/api/headers/header-setcookie.any.js
Comment (jimmywarting) [inline fetch/api/headers/header-setcookie.any.js:141]: shouldn't this and several others (where you compare arrays) use `assert_array_equals` instead?

File: fetch/api/headers/header-setcookie.any.js
Comment (jimmywarting) [inline fetch/api/headers/header-setcookie.any.js:163]: Same here... not a 2D array...

**Comment by lucacasonato** [comment]
Thanks @jimmywarting - fixed the tests.

I implemented `getSetCookie` for Deno in this PR: https://github.com/denoland/deno/pull/13542. All added WPT tests pass.

File: fetch/api/headers/header-setcookie.any.js
Comment (andreubotella) [inline fetch/api/headers/header-setcookie.any.js:137]: It'd be good to have a test for `append` as well.

**Comment by andreubotella** [comment]
For the record, the Webkit implementation will have to treat `Set-Cookie` as different from all other headers, special-casing it in every single operation. With that in mind, it would be good to also add some smoke tests for `delete()` and `has()`. But those don't need to block this PR, I can add them in after it lands.

File: fetch/api/headers/header-setcookie.any.js
Comment (annevk) [inline fetch/api/headers/header-setcookie.any.js:31]: Why do these not have asserts?

File: fetch/api/headers/header-setcookie.any.js
Comment (annevk) [inline fetch/api/headers/header-setcookie.any.js:141]:
```suggestion
  assert_true(iterator.next().done);
```

**Review by annevk — COMMENTED** [review]
The only thing that comes to mind is there's no coverage for `Headers` objects in conjunction with `Request` or `Response` objects. Perhaps we should wrap all these tests in a loop so we can test all variants or are implementations not specialized enough to warrant any of that?

**Review by annevk — APPROVED** [review]
I meant to approve, though I'd like to see the feedback addressed in some manner before this lands together with the spec change.

File: fetch/api/headers/header-setcookie.any.js
Comment (andreubotella) [reply fetch/api/headers/header-setcookie.any.js:31]: I suppose they're meant to test that the constructor doesn't throw, but the rest of tests also do that. I'd be fine with, for example, adding asserts to the first two, and removing the third one since it's redundant with some of the other tests that also use `new Headers(headerList)`.

File: fetch/api/headers/header-setcookie.any.js
Comment (MayyaSunil) [inline fetch/api/headers/header-setcookie.any.js:5]: We could have tests where set-cookie headers have empty values. 
From the draft spec, I understand that empty values are allowed?

**Comment by andreubotella** [comment]
> The only thing that comes to mind is there's no coverage for `Headers` objects in conjunction with `Request` or `Response` objects. Perhaps we should wrap all these tests in a loop so we can test all variants or are implementations not specialized enough to warrant any of that?

#34424 added tests for `Request`. [`fetch/api/cors/cors-expose-star.sub.any.js`](https://github.com/web-platform-tests/wpt/blob/master/fetch/api/cors/cors-expose-star.sub.any.js) tests that responses returned from `fetch()` don't include the `Set-Cookie` header. There didn't seem to be tests about adding `Set-Cookie` headers to `Response`s though, so I added one.

File: fetch/api/headers/header-setcookie.any.js
Comment (annevk) [reply fetch/api/headers/header-setcookie.any.js:5]: There is such a test further down.

**Pull request merged**
